aten/src/ATen/native/cuda/Sort.cu

```diff
@@ -46,6 +50,306 @@ bool should_use_th_sort(const Tensor &self, int64_t dim) {
+    int64_t numThreads = sliceSize;
+    if (numThreads > maxThreads) {
+      numThreads = maxThreads;
+    }
+
+    dim3 block(numThreads);
+#define FILL_INDEX(T, DIM)                                         \
+    fillSliceWithIndex_kernel<T, DIM>                                     \
+      <<<grid, block, 0, at::cuda::getCurrentCUDAStream()>>>(      \
+        info, numSlices, sliceSize, info.strides[collapseDim]);    \
+    C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+
+    if (at::cuda::detail::canUse32BitIndexMath(t)) {
+      at::cuda::detail::TensorInfo<int64_t, unsigned int> info =
+        at::cuda::detail::getTensorInfo<int64_t, uint32_t>(t);
+      if (!t.dim()) {
+        info.dims = 1;
+        info.strides[0] = 1;
+      }
+      info.reduceDim(dim);
+      int collapseDim = info.collapseDims(dim);
+      if (info.isContiguous()) {
+        FILL_INDEX(unsigned int, -2);
+      } else {
+        if (info.dims == 1) {
+          FILL_INDEX(unsigned int, 1);
+        } else if (info.dims == 2) {
+          FILL_INDEX(unsigned int, 2);
+        } else {
+          FILL_INDEX(unsigned int, -1);
+        }
+      }
+    } else {
+      at::cuda::detail::TensorInfo<int64_t, uint64_t> info =
+        at::cuda::detail::getTensorInfo<int64_t, uint64_t>(t);
+      if (!t.dim()) {
+        info.dims = 1;
+        info.strides[0] = 1;
+      }
+      info.reduceDim(dim);
+      int collapseDim = info.collapseDims(dim);
+
+      // catch-all implementation
+      FILL_INDEX(uint64_t, -1);
+    }
+#undef FILL_INDEX
+  }
+}
+
+std::tuple<Tensor &,Tensor &> small_sort_out_cuda(const Tensor &self,
+                                                  c10::optional<bool> stable,
+                                                  Tensor &values,
+                                                  Tensor &indices,
+                                                  int dim, bool order) {
+  TORCH_INTERNAL_ASSERT(stable.has_value(), "small_sort(): c10::optional<bool> for stable has to have value.");
+  TORCH_CHECK(!stable.value(), "stable=True is not implemented on CUDA yet.");
+
+  // from thc: sorted->values, indices->indices, input->self
+  TensorArg self_arg{self, "self", 1}, values_arg{values, "values", 2}, indices_arg{indices, "indices", 3};
+  checkAllSameGPU("small_sort", {self_arg, values_arg, indices_arg});
+  dim = at::maybe_wrap_dim(dim, self);
+  int64_t dims = self.dim();
+  dims = dims == 0 ? 1 : dims;
+  TORCH_CHECK(dims <= MAX_DIMS, "self tensor has too many dimensions");
+  dims = values.dim();
+  dims = dims == 0 ? 1 : dims;
+  TORCH_CHECK(dims <= MAX_DIMS, "values tensor has too many dimensions");
+  dims = indices.dim();
+  dims = dims == 0 ? 1 : dims;
+  TORCH_CHECK(dims <= MAX_DIMS, "indices tensor has too many dimensions");
+
+  if (!values.defined()) {
+    values = at::empty_like(self, self.options());
+  }
+  if (!indices.defined()) {
+    indices = at::empty_like(self, self.options().dtype(kLong));
+  }
```

Comment:
Instead of repeating all the checking logic here, could you refactor your code to reuse the checks in `sort_out_stable_cuda`?